### ISSUE TITLE
[pose-generation] Add state-keyed low-discrepancy candidate streams and multi-root replicas

### ISSUE THREAD

**Issue by JanDuchscherer104** [JanDuchscherer104/ARIA-NBV#68]
Parent: #67

## Outcome

Replace per-call seed resetting with a deterministic **state-keyed random-stream contract**, add optional low-discrepancy candidate designs, and generate controlled root/candidate replicas so that the rollout corpus varies reconstruction stage and action support without sacrificing exact reproducibility.

This is the highest-ROI diversity patch because it changes neither the oracle objective nor the action semantics; it removes artificial correlations caused by repeatedly drawing the same normalized shell.

## Current repository state

At the PR #62 head:

- `CandidateViewGenerator.generate()` enters `_maybe_seed(self.config.seed)` on every call.
- `CandidateMixtureViewGenerator._component_config()` derives each component seed as `base.seed + component_index`.
- Every successor node therefore reuses the same component-specific random template, transformed by the new reference pose and target.
- The rollout selection RNG is also initialized from the same recipe seed for every work unit.
- `max_resamples` exists in configuration, but the current generation path does not implement a bounded multi-round refill loop.
- A source task normally uses one reference/root frame even though the generator supports `reference_frame_index`.

Relevant owners:

- `aria_nbv/aria_nbv/pose_generation/candidate_generation.py`
- `aria_nbv/aria_nbv/pose_generation/positional_sampling.py`
- `aria_nbv/aria_nbv/pose_generation/candidate_mixture.py`
- `aria_nbv/aria_nbv/rollouts/replay/engine.py`
- `aria_nbv/aria_nbv/oracle/pipelines/rollout_dataset.py`
- `aria_nbv/aria_nbv/rollouts/zarr_store.py`

## Theory

Determinism should be a pure function of the physical rollout state and a named replica, not a global RNG reset:

```math
\xi_{t,k,r} = H(
  \text{generation-contract},
  \text{campaign-seed},
  \text{sample},
  \text{target},
  \text{root},
  \text{selected-history},
  t,
  k,
  r,
  \text{draw-round}
).
```

Two policies evaluating the **same state** should receive the same candidate table. Two different roots, histories, candidate replicas, components, or draw rounds should receive different deterministic streams.

Low-discrepancy designs are useful because the candidate budget is small. PyTorch describes Sobol samples as “low discrepancy quasi-random sequences”; they cover a bounded parameter cube more evenly than IID draws at equal sample count.

Power Spherical remains useful for smooth forward bias because it has “support on the hyper-sphere, symmetry about its mean direction parameter.” Do not silently claim that a Sobol yaw/elevation design has the exact Power-Spherical density; expose them as distinct sampling strategies.

## Proposed data contract

Add immutable typed keys rather than passing bare integer seeds:

```python
@dataclass(frozen=True, slots=True)
class CandidateSamplingKey:
    generation_contract: str
    campaign_seed: int
    sample_key: str
    target_id: str
    root_id: str
    selected_shell_history: tuple[int, ...]
    step_index: int
    component_name: str
    candidate_replica_id: int
    draw_round: int

    def seed64(self) -> int: ...
```

Add persisted per-step/per-component provenance:

```text
sampling_key_hash
candidate_replica_id
root_variant_id
root_frame_index / root_time_ns
sampling_strategy
sequence_type = iid | sobol_scrambled | stratified
sequence_scramble_seed
sequence_start_index
component_seed
selection_seed
draw_round
```

Do not store the full sensitive key redundantly in every candidate row; bind it once at step/component level and retain stable row indices.

### 1. Remove global RNG dependence

- Pass an explicit `torch.Generator` or typed random-stream object into position and orientation samplers.
- Do not call `torch.manual_seed` or `torch.cuda.manual_seed_all` inside ordinary candidate generation.
- Keep the caller's global CPU/CUDA RNG states unchanged.
- Derive independent substreams for position, orientation, family allocation, refill, and rollout selection.

### 2. Add low-discrepancy bounded designs

Implement an explicit strategy such as `SOBOL_CAPPED_LOCAL` using `torch.quasirandom.SobolEngine` over normalized dimensions:

```text
u_r      -> radius distribution
u_az     -> bounded azimuth
u_el     -> uniform sin(elevation)
u_yaw    -> bounded local yaw
u_pitch  -> bounded local pitch
optional u_roll -> small roll robustness ablation
```

Requirements:

- use scrambling with the state-keyed seed;
- preserve candidate row order and sequence index;
- test discrepancy/coverage empirically rather than assuming superiority;
- retain the existing Power-Spherical sampler as a separate state-keyed option;
- do not fall back silently from one strategy to another.

### 3. Generate candidate replicas

Add manifest-resolved `candidate_replica_count`. A replica changes only the proposal stream; it must not change target identity, oracle settings, or scene split.

Recommended pilot:

```text
candidate_replica_count = 2
```

Use one frozen replica for matched confirmatory policy comparison. Additional replicas expand training support and proposal-quality analysis.

### 4. Diversify root states

Add deterministic root variants sampled from the logged prefix, for example:

```text
root_strategy = quantile_frames
root_quantiles = [0.35, 0.70, 1.00]
```

or an explicit frame list in the source manifest. Each root must carry exact frame/time/trajectory lineage. Root variants from one scene remain in one split and are aggregated within scene.

The purpose is to vary:

- reconstruction stage and root target error;
- target bearing and distance;
- observed target support;
- local clearance and candidate survival;
- stage-dependent RRI distribution.

### 5. Keep matched-policy support exact

For a fixed `(source, target, root, replica, state-history)`:

- random-valid, oracle-greedy, learned-myopic, and learned finite-horizon policies must receive byte-equivalent candidate tables;
- policy identity must not enter the candidate sampling key;
- selected history does enter the key because it changes the physical successor state.

## Acceptance criteria

- [ ] Same `CandidateSamplingKey` produces bitwise-identical full-shell poses, masks, provenance, and row order on the supported deterministic backend.
- [ ] Changing only root, history, component, replica, or draw round changes the stream while preserving configured marginals.
- [ ] Candidate generation does not mutate global CPU or CUDA RNG state.
- [ ] Two policies at the same physical state receive identical candidate tables.
- [ ] At least one Sobol/stratified strategy is implemented and named separately from Power Spherical.
- [ ] Root variants and candidate replicas are persisted and remain scene-split consistent.
- [ ] Store readers expose root/replica identity without making it an oracle-only model shortcut.
- [ ] Proposal diagnostics compare IID Power-Spherical, state-keyed IID, and low-discrepancy designs on equal `N_q`.
- [ ] Existing single-root/single-replica stores remain readable under explicit legacy provenance.

## Required tests

- deterministic replay across process restart;
- different-history seed separation;
- matched-policy candidate equality;
- no global-RNG side effect;
- CPU/CUDA provenance and supported determinism behavior;
- root-frame lineage and scene-split containment;
- sequence-index stability under component concatenation;
- row-shuffle-independent physical identity;
- distribution sanity for radius, azimuth, `sin(elevation)`, yaw, and pitch;
- duplicate-rate and pairwise-distance comparison against the current seed-reset baseline.

## Evidence to report before promotion

For each strategy and root cohort:

- valid count and zero-action rate;
- duplicate/near-duplicate rate in SE(3);
- pairwise translation, yaw, and target-bearing separation;
- target-framing distribution;
- best-of-60 and top-5 oracle target gain;
- family survival and selected-family share;
- renderer/scorer time per effective valid row;
- within-scene versus between-scene variation.

## External sources

- PyTorch `SobolEngine`: https://docs.pytorch.org/docs/stable/generated/torch.quasirandom.SobolEngine.html
- De Cao & Aziz, *The Power Spherical distribution*: https://arxiv.org/abs/2006.04437

## Non-goals

- This issue does not learn a proposal distribution.
- It does not alter RRI, target cropping, validity semantics, or the canonical candidate count.
- It does not use policy identity to decorrelate candidate tables.
- It does not make multi-root replicas independent statistical units; primary aggregation remains scene-level.

**Comment by JanDuchscherer104** [comment]
## Research enrichment — round 2: deterministic scope, direct spherical designs, and correct pose measures

### 1. Define the scope of reproducibility precisely

PyTorch's official reproducibility note states that completely reproducible results are not guaranteed across releases, platforms, or CPU/GPU execution:

https://pytorch.org/docs/stable/notes/randomness.html

Therefore replace an unconditional “bitwise deterministic” promise with two explicit levels:

```text
REPLAY_EXACT
  exact environment lock, device/backend, dependency versions, deterministic-algorithm settings

REPLAY_TOLERANT
  different supported hardware/backend; stable identities and geometry within frozen numeric tolerances
```

Persist the environment/lock hash, Torch/CUDA/PyTorch3D versions, deterministic-algorithm flags, device/backend, and whether exact or tolerant replay is claimed. Cross-release differences must not be mistaken for candidate-policy changes.

### 2. Compare Sobol mappings with direct low-discrepancy designs on `S²`

A Sobol sequence is naturally defined on a unit cube. Mapping it to azimuth/elevation can preserve a desired measure only if the transform is explicit—e.g. uniform azimuth and uniform `sin(elevation)` for an area-uniform spherical band.

Add a direct spherical design control:

- Marques et al., *Spherical Fibonacci Point Sets for Illumination Integrals*: https://doi.org/10.1111/cgf.12190
- Keinert et al., *Spherical Fibonacci Mapping*: https://doi.org/10.1145/2816795.2818131

Spherical Fibonacci points provide deterministic, well-separated near-uniform directions without first drawing IID noise. Use them for the broad/uniform component and compare against:

```text
state-keyed IID Power Spherical
scrambled Sobol transformed to the configured band
spherical Fibonacci / stratified angular banks
```

Do not describe capped/remapped samples as Power-Spherical distributed after a nonlinear cap transform. Persist `base_measure`, `cap_transform`, and `resolved_design_measure` separately.

### 3. Treat the pose measure as part of the action contract

Uniform Euler angles do not imply a uniform rotation distribution. Shoemake's classic construction and Kuffner's SE(3) planning note are useful references:

- Shoemake, *Uniform Random Rotations*: https://www.cs.cmu.edu/~kiranb/animation/p245-shoemake.pdf
- Kuffner, *Effective Sampling and Distance Metrics for 3D Rigid Body Path Planning*: https://www.ri.cmu.edu/pub_files/pub4/kuffner_james_2004_2/kuffner_james_2004_2.pdf

ARIA-NBV uses a reduced 5-DoF action, so it need not sample Haar-uniform `SO(3)`. It must nevertheless define the intended measure over center, yaw, and pitch rather than relying on independent-angle intuition. For each strategy report marginal and joint distributions in physically meaningful coordinates.

### 4. Use versioned, collision-resistant stream derivation

Derive substreams with a versioned cryptographic digest over canonical bytes, not Python's process-dependent `hash()`:

```text
seed_derivation_version
canonical state-key serialization
full digest
truncated uint64 seed
collision test over rendered campaign keys
```

Use named substreams (`center`, `gaze`, `family-allocation`, `refill`, `selection`) so adding one random draw to orientation code does not silently change every downstream decision.

### 5. Add support-design diagnostics beyond marginal histograms

For every design compare:

```text
spherical-cap discrepancy / equal-area bin occupancy
nearest-neighbor angular distance
pair-correlation / near-duplicate rate
minimum and quantile SE(3) separation
best-of-K oracle gain versus K
stability across replicas
```

The promotion criterion should prefer better coverage and oracle support under equal candidate and attempt budgets—not low discrepancy in isolation.

### Additional acceptance criteria

- [ ] Exact and tolerance-based reproducibility claims are distinguished and environment-bound.
- [ ] A direct spherical Fibonacci/stratified control is compared with Sobol and Power Spherical.
- [ ] Every sampler records the intended base measure and cap transform.
- [ ] Seed derivation is canonical, versioned, collision-tested, and independent of Python hash randomization.
- [ ] Named substreams prevent unrelated implementation changes from shifting all samples.
- [ ] Equal-area directional coverage and nearest-neighbor separation are part of the evidence artifact.
- [ ] Candidate tables remain matched across policies even after implementation of replicas and low-discrepancy designs.

**Comment by JanDuchscherer104** [comment]
## Handoff — first implementation slice

**Mission:** remove repeated-shell RNG correlation and make candidate sampling reproducibly state-dependent before adding new semantic pose families.

### Implement now
Create a typed state/sample key and derive independent deterministic substreams from at least:
- campaign/base seed;
- source/sample identity;
- target-task identity;
- root/reference-frame identity;
- selected-history or canonical state identity;
- rollout step;
- candidate component/family;
- candidate-set replica;
- draw/refill round.

**Do not include policy/recipe name** in the physical candidate-shell seed: matched policies reaching the same physical state must see the same shell.

Replace repeated per-call RNG resets with explicit local generators/substreams. Add a low-discrepancy implementation (prefer scrambled Sobol for the first control) for the continuous component parameters while retaining the present sampler as the matched baseline.

Persist enough provenance to reproduce each stream/design: sampling-contract version, resolved key/digest, substream role, design type, sequence/scramble index or seed, component, replica, and draw round.

### Tests required
- same physical state + different policy recipe ⇒ same candidate shell;
- same key after process restart ⇒ same shell within the declared reproducibility tolerance;
- changing target/root/step/replica ⇒ changes the intended stream;
- sampler calls do not mutate global Torch/Python/NumPy RNG state;
- family counts, masks, and stable row identities remain aligned;
- Sobol/design sequence has no accidental duplicated rows before geometric pruning.

### Non-goals
- No orbit/standoff/turn-in-place families; #69 owns semantic family expansion.
- No adaptive schedule; #70 owns state-conditioned allocation.
- No reservoir/refill policy beyond the provenance hooks needed by #71.
- Do not promote Sobol or any changed support distribution as canonical in this PR.

### Stop/escalate when
The patch changes physical support, family semantics, or policy comparison matching beyond the declared sampler-control experiment. Such changes require a versioned candidate contract and #73 evidence.

### Done when
Candidate randomness is local, state-keyed, independently replayable, policy-neutral for matched states, fully provenance-bound, and both IID-control and low-discrepancy streams are available for paired benchmarking.
